src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs

```diff
@@ -104,14 +125,97 @@ private void RecordDependencyGraphFromFile(PnpmYaml yaml, ISingleFileComponentRe
+    {
+        // There may be multiple instance of the same package (even at the same version) in pnpm differentiated by other aspects of the pnpm dependency path.
+        // Therefor all DetectedComponents are tracked by the same full string pnpm uses, the pnpm dependency path, which is used as the key in this dictionary.
+        // Some documentation about pnpm dependency paths can be found at https://github.com/pnpm/spec/blob/master/dependency-path.md.
+        var components = new Dictionary<string, (DetectedComponent, Package)>();
+
+        // Create a component for every package referenced in the lock file.
+        // This includes all directly and transitively referenced dependencies.
+        foreach (var (pnpmDependencyPath, package) in yaml.packages ?? Enumerable.Empty<KeyValuePair<string, Package>>())
+        {
+            // Ignore "file:" as these are local packages.
+            // Such local packages should only be referenced at the top level (via ProcessDependencyList) which also skips them or from other local packages (which this skips).
+            // There should be no cases where a non-local package references a local package, so skipping them here should not result in failed lookups below when adding all the graph references.
+            if (pnpmDependencyPath.StartsWith("file:"))
+            {
+                continue;
+            }
+
+            var parentDetectedComponent = PnpmParsingUtilities.CreateDetectedComponentFromPnpmPathV6(pnpmDependencyPath: pnpmDependencyPath);
+            components.Add(pnpmDependencyPath, (parentDetectedComponent, package));
+
+            // Register the component.
+            // It should get registered again with with additional information (what depended on it) later,
+            // but registering it now ensures nothing is missed due to a limitation in dependency traversal
+            // like skipping local dependencies which might have transitively depended on this.
+            singleFileComponentRecorder.RegisterUsage(parentDetectedComponent, isDevelopmentDependency: PnpmParsingUtilities.IsPnpmPackageDevDependency(package));
+        }
+
+        // Now that the `components` dictionary is populated, make a second pass registering all the dependency edges in the graph.
+        foreach (var (_, (component, package)) in components)
+        {
+            foreach (var (name, version) in package.dependencies ?? Enumerable.Empty<KeyValuePair<string, string>>())
+            {
+                var pnpmDependencyPath = PnpmParsingUtilities.ReconstructPnpmDependencyPathV6(name, version);
+
+                // If this lookup fails, then pnpmDependencyPath was either parsed incorrectly or constructed incorrectly.
+                var (referenced, _) = components[pnpmDependencyPath];
+
+                singleFileComponentRecorder.RegisterUsage(referenced, parentComponentId: component.Component.Id, isExplicitReferencedDependency: false);
```

Comment:
Its unclear to me if I need to respecify dev-dependency-ness here. Would doing so indicate this dependency edge is a dev dependency, or and I overwriting the dev-dependency ness of the target component with the default by not specifying it here?

Reply:
Our component merging logic will do the following where if any detection is classified as non-dev it will categorize the component as non-dev.

First Detection Dev Dep | Second Detection Dev Dep | Final Result
false                             | false                              | false
true                              | true                               | true
true                              | false                              | false

Reply:
When passing null for isDevelopmentDependency whatever the previous value is will stay.